monogen/monogen_test.go

```diff
@@ -0,0 +1,69 @@
+package monogen_test
+
+import (
+	"context"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"testing"
+
+	bftclient "github.com/cometbft/cometbft/rpc/client/http"
+	"github.com/polymerdao/monomer/e2e/url"
+	"github.com/polymerdao/monomer/monogen"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGenerate(t *testing.T) {
+	const appName = "testapp"
+	rootDirPath := t.TempDir()
+	appDirPath := filepath.Join(rootDirPath, appName)
+	require.NoError(t, monogen.Generate(context.Background(), appDirPath, "github.com/test/"+appName, "test", true))
+
+	initCommand := func(cmd *exec.Cmd) *exec.Cmd {
+		cmd.Dir = appDirPath
+		// Set a different home directory to avoid cluttering the home directory of the person running the script.
+		// Ignite prefixes environment variables with the lowercase app name instead of uppercase for some reason.
+		cmd.Env = append(os.Environ(), appName+"_HOME="+rootDirPath)
+		logger := logWriter(t.Log)
+		cmd.Stdout = logger
+		cmd.Stderr = logger
+		return cmd
+	}
+
+	// Run monogen.sh.
+	scriptPath, err := filepath.Abs("monogen.sh")
+	require.NoError(t, err)
+	require.NoError(t, initCommand(exec.Command(scriptPath)).Run())
+
+	// Start the app.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	cmd := initCommand(exec.CommandContext(ctx, filepath.Join(appDirPath, appName+"d"), "monomer", "start", "--minimum-gas-prices", "0.0001stake"))
+	require.NoError(t, cmd.Start())
+	defer func() {
+		cancel()
+		// TODO: handle this error.
+		// It is kind of harmless, just `signal: killed` or `signal: terminated`,
+		// but it would be nice to be able to say `require.NoError` here.
+		_ = cmd.Wait()
+	}()
```

Comment:
Why is this `defer func` necessary? Isn't the context already properly canceled with the `defer cancel()` call above?

Reply:
`defer`s are executed in reverse order.

`Wait` will only finish if `cancel` is called first.

This means that `cancel` is called twice on the happy path.

The first `defer cancel()` just ensures `cancel` is called on even the strangest unhappy paths.